### PR TITLE
Implement support for XamlDirect.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Add new transforms and emitters to enable producing code that uses XamlDirect when possible and falling back to DependencyObjects when needed. Update our test library's XAML to cover more cases.

Also implement support for old-style bindings on dependency properties since I was already in that area of the code.

### CODE REVIEW

File: WinUIXamlCompiler/WinUIXamlCompilerExtensions.cs [hunk 1]
```diff
@@ -19,7 +21,31 @@ void InsertAfter<T>(params IXamlAstTransformer[] t)
 
             void InsertBefore<T>(params IXamlAstTransformer[] t)
                 => compiler.Transformers.InsertRange(compiler.Transformers.FindIndex(x => x is T), t);
+
+            InsertAfter<ResolveContentPropertyTransformer>(
+                new XamlDirectPropertyResolver(),
```
Comment (stevenbrix) [inline WinUIXamlCompiler/WinUIXamlCompilerExtensions.cs:26]: how do you know where to insert certain resolvers at?

File: WinUIXamlCompiler/Transforms/XamlDirectPropertyResolver.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            if (node is XamlAstClrProperty prop)
+            {
+                if (prop.DeclaringType.Assembly == context.GetWinUITypes().WinUIControlsAssembly)
+                {
+                    IXamlField propertyIndexMaybe = context.GetWinUITypes().XamlPropertyIndex.Fields.FirstOrDefault(f => f.Name == $"{prop.DeclaringType.Name}_{prop.Name}");
+                    if (propertyIndexMaybe != null)
+                    {
+                        prop.Setters.Insert(0, new XamlDirectSetter(context.GetWinUITypes(), prop.Getter.ReturnType, prop.DeclaringType, propertyIndexMaybe));
+                        foreach (var adder in XamlTransformHelpers.FindPossibleAdders(context, prop.Getter.ReturnType))
```
Comment (stevenbrix) [inline WinUIXamlCompiler/Transforms/XamlDirectPropertyResolver.cs:25]: what does it mean to "FindPossibleAdders"?

**Review by stevenbrix — APPROVED** [review]

File: WinUIXamlCompiler/Transforms/XamlDirectPropertyResolver.cs (hunk 2 above)
Comment (jkoritzinsky) [reply WinUIXamlCompiler/Transforms/XamlDirectPropertyResolver.cs:25]: This finds add methods on the type of the property (like `UIElementCollection.Add`).

File: WinUIXamlCompiler/WinUIXamlCompilerExtensions.cs (hunk 1 above)
Comment (jkoritzinsky) [reply WinUIXamlCompiler/WinUIXamlCompilerExtensions.cs:26]: Mostly experience. The idea here was "I want to do something after I've resolved all properties but before I've figured out how I can assign the values" so I put it after the last phase that resolves properties. I could have also put it before the transform that resolves assignments (which IIRC would be the same place in the pipeline).

Comment (jkoritzinsky) [reply WinUIXamlCompiler/WinUIXamlCompilerExtensions.cs:26]: There's a few phases that are pretty common ones to put things before/after. Take a look at Avalonia's compiler for some examples.

**Pull request merged**